code/game/machinery/PDApainter.dm

```diff
@@ -84,42 +98,88 @@
 			stat |= BROKEN
 			update_icon()
 
-/obj/machinery/pdapainter/attack_hand(mob/user as mob)
+/obj/machinery/pdapainter/attack_hand(mob/user)
 	if(..())
-		return 1
+		return TRUE
 
-	src.add_fingerprint(user)
+	ui_interact(user)
 
-	if(storedpda)
-		var/obj/item/pda/P
-		P = input(user, "Select your color!", "PDA Painting") as null|anything in colorlist
-		if(!P)
-			return
-		if(!in_range(src, user))
-			return
+/obj/machinery/pdapainter/ui_interact(mob/user, ui_key, datum/tgui/ui, force_open, datum/tgui/master_ui, datum/ui_state/state)
+	ui = SStgui.try_update_ui(user, src, ui_key, ui, force_open)
+	if(!ui)
+		ui = new(user, src, ui_key, "PdaPainter", "PDA Painter", 510, 510)
+		ui.open()
 
-		storedpda.icon_state = P
-		storedpda.desc = colorlist[P]
+/obj/machinery/pdapainter/ui_data(mob/user)
+	var/list/data = list()
+	data["has_pda"] = storedpda ? TRUE : FALSE
+	if(storedpda)
+		var/icon/pda_sprite = icon(storedpda.icon, storedpda.icon_state, frame = 1)
+		data["current_appearance"] = icon2base64(pda_sprite)
```

Comment:
could you cache this? What I mean by this is storing the `icon2base64` output into a variable stored on the PDApainter machine. You can update this var whenever a pda is inserted, which means it won't have to recalculate the icon/img every time `ui_data()` is called.

I know it's a bit of an ask but it will be more performative, everything else looks good however.

Reply:
cached